Guide/src/reference/architecture/openhcl_servicing.md

```diff
@@ -0,0 +1,258 @@
+- **Correlation ID**: For tracing across the servicing operation
+
+#### Device-Specific State
+- **NVMe State**: NVMe manager and driver state (when keepalive is enabled)
+- **DMA Manager State**: Private and shared pool allocations
+- **VMBus Client State**: Client-side VMBus connection state
+- **MANA State**: Microsoft Azure Network Adapter state
+
+### 3. Component Shutdown
+
+After state is saved, components are shut down:
+
+#### Without Keepalive
+- NVMe devices are cleanly shut down
+- Device handles are closed
+- DMA allocations are released
+- VFIO device handles are dropped (causing device reset)
+
+#### With Keepalive (NVMe)
+- NVMe devices remain connected (CC.EN=1)
+- VFIO device handles are kept open (preventing reset)
+- DMA buffers in the private pool are preserved
+- Device maintains its operational state
+
+### 4. Host Servicing
+
+While OpenHCL is stopped:
+- The host can update the OpenHCL binary
+- The host can update the OpenHCL kernel
+- The host can update the IGVM file
+- VM guest state in VTL0 continues to be preserved by the host
+
+### 5. OpenHCL Restart
+
+The host restarts OpenHCL with the new version:
+- New OpenHCL instance loads
+- Saved state is provided as input
+- Memory layout and resources are restored
+
+### 6. State Restoration
+
+OpenHCL restores components in order:
+
+1. **DMA Manager Restoration**: Private and shared pools are restored first
+2. **NVMe Manager Restoration**: NVMe devices reconnect to their saved state
+3. **Device Restoration**: VFIO devices are opened and reconnected
+4. **State Unit Restoration**: VM devices and emulation state is restored
+5. **VM Resumption**: The guest VM resumes execution
+
+## NVMe Keepalive
+
+NVMe keepalive is a key feature that allows NVMe devices to remain operational during servicing:
+
+### Requirements
+
+NVMe keepalive requires all of the following:
+
+1. **Private Pool Availability**: The DMA manager must have private pool ranges configured
+2. **Host Support**: The host must support keepalive operations
+3. **Configuration**: `nvme_keep_alive` must be enabled in OpenHCL configuration
+4. **Save/Restore Support**: Must be enabled when creating the NVMe manager
+
+### How It Works
+
+When keepalive is enabled:
+
+1. **Persistent DMA Allocations**: NVMe driver uses the private pool for all DMA buffers
+2. **State Preservation**: 
+   - NVMe driver saves queue states, registers, and namespace information
+   - DMA manager saves private pool allocation metadata
+   - VFIO keeps device handles open
+3. **Device Stays Connected**: The NVMe controller remains enabled (CC.EN=1)
+4. **Restoration**:
+   - Private pool allocations are restored
+   - VFIO device is reconnected with persistent DMA clients
+   - NVMe driver restores queue state and resumes I/O operations
+
+### Benefits
+
+- **Minimal Downtime**: No device reset or reinitialization required
+- **No I/O Interruption**: Pending I/O operations can complete
```

Comment:
This may be true. Make note that IO is saved at some point in the save process. Point out that IOs that are not "hooked up", meaning that while IO may complete, it will still be re-issued.